### PR TITLE
AnneJu optional 2: share tasks among users

### PR DESCRIPTION
**Pull request by anneju**
#### Description
- Want to allow multiple people to view and edit the same task
- Show task creator
- ref: [Option Requirement 2: Want to be able to share tasks among users](https://github.com/Fablic/training/blob/develop/steps_en.md#option-requirement-2-want-to-be-able-to-share-tasks-among-users)

#### Stories
- [x] Add model EditableTaskUser to store shared task and user association
- [x] Manage to CRUD in controller
- [x] Write Specs

#### Images
`Mine` tasks
Screen Shot 2023-01-19 at 13 33 39

`Others` tasks
Screen Shot 2023-01-19 at 13 35 48

#### Related RSpecs
```
rspec spec/models/editable_task_user_spec.rb spec/system/tasks_spec.rb:179:272:286:438
```

### CODE REVIEW

File: myapp/app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -65,12 +66,12 @@ def complete
...
   def task_params
-    params.require(:task).permit(:title, :description, :due_date, :tag_list)
+    params.require(:task).permit(:title, :description, :due_date, :tag_list, :editable_user_list)
   end
 
   def set_task
-    @task = Task.find_by(id: params[:id], user: Current.user)
-    raise ActiveRecord::RecordNotFound if @task.nil?
+    @task = Task.find_by(id: params[:id])
+    raise ActiveRecord::RecordNotFound if @task.nil? || (@task.user_id != Current.user.id && !Current.user.id.in?(@task.editable_users.pluck(:id)))
```
Comment (anneju) [inline myapp/app/controllers/tasks_controller.rb:74]: only allow the author or editable_users to access

File: myapp/app/models/task.rb [hunk 2]
```diff
@@ -74,4 +77,16 @@ def tag_list=(names)
...
     Tag.find_by(name: name)&.tasks || Task.none
   end
+
+  def editable_user_list
+    editable_users.map(&:email).join(', ')
+  end
+
+  def editable_user_list=(emails)
+    self.editable_users = emails.split(',').map do |email|
+      user_id = User.find_by!(email: email.strip).id
```
Comment (anneju) [inline myapp/app/models/task.rb]: I know it's not ideal if user input emails that no user has, but just to make it simple here as a better solution will involve more complicated FE works. 😅

**Review by sangjin-oh — COMMENTED** [review]
It seems writer's task is included in others's tasks list.



Is it ok?

**Comment by anneju** [comment]
@sangjin-oh Thanks for spotting that! Yeah, it can be improved as I didn't have a good filter when adding the author as an `editable user`. I have enhanced with 0c8291d166be68526578061aad0c8d2138185023 and add RSpecs to make sure it does not happen~

Screen Shot 2023-01-30 at 14 45 56

**Review by sangjin-oh — APPROVED** [review]
LGTM!

LGTM

**Pull request merged**